### ISSUE TITLE
Accessibility permission issue

### ISSUE THREAD

**Issue by AlainVagner** [qpdf/qpdf#526]
Hi, I have an issue with [a pdf file](https://gouvernement.lu/dam-assets/fr/publications/rapport-activite/minist-affaires-etrangeres-europeennes/armee/2004-rapport-activite-armee/2004-rapport-activite-armee.pdf).

According to `qpdf --check 2004-rapport-activite-armee.pdf` the permission to extract for accessibility is not allowed.
However, according to Acrobat, the permission is indeed allowed.
<img width="283" alt="Screenshot 2021-04-29 at 21 12 00" src="https://user-images.githubusercontent.com/16536731/116605407-ab741280-a92f-11eb-851f-d699f0a8fd99.png">

**Comment by jberkenbilt** [comment]
Newer versions of the PDF spec deprecated disallowing accessibility, and modern, conforming readers are supposed to ignore that bit. Someone figured out that it's not okay to block accessibility in a document because it discriminates against people who require the accessibility features. So this behavior is as expected.

**Comment by AlainVagner** [comment]
Thanks for your feedback, I was unaware of this deprecation in the PDF spec. Unfortunately the latest version of Acrobat on Windows does not seem to be a conforming reader. This flag is still enforced, I just tested this with the Jaws and NVDA screen readers in their latest versions.
When this flag is set to false, a screen reader will only tell that the document is "secured" and won't read the content at all.
I am implementing a PDF accessibility checking tool based on pikepdf, and trying to implement the check of this flag (the [Matterhorn Protocol](https://www.pdfa.org/resource/the-matterhorn-protocol/), which was updated a few days ago still mentions this test). It looks like I will need to read the P field by myself. No big deal, I just wanted to be sure.

**Comment by jberkenbilt** [comment]
For your reference and in case anyone asks, in the PDF 2.0 spec, (ISO/DIS 32000-2 is the copy I have; I have not forked over the money to purchase the latest version yet but probably will), section 7.6.3.2 "Standard encryption dictionary", table 22, bit position 5: "Copy or otherwise extract text and graphics from the document. However, for the limited purpose of providing this content to assistive technology, a PDF reader should behave as if this bit was set to 1. Note: ISO 32000-1 had this option restricted by bit 10, for accessibility, but that exception has been deprecated.", and bit position 10: "Not used. This bit was previously used to determine whether content could be extracted for the purposes of accessibility, however, that restriction has been deprecated. PDF readers shall ignore this bit and PDF writers shall always set this bit to 1 to ensure compatibility with PDF readers that followed earlier specifications."

Note that qpdf still allows that bit to be set or cleared. This is because an important purpose of qpdf is to allow creation of all sorts of files for testing purposes. If you wanted, for example, to create a file that disallowed accessibility to figure out which readers obey they rule to ignore the bit, you could use qpdf to do this. You can also use qpdf to change the encryption on a file, though doing so would invalidate any digital signatures on the file.

**Comment by AlainVagner** [comment]
Thanks for the info, indeed I don't have access to the PDF 2.0 spec. I am just wondering how should readers interpret the flags of files produced with previous versions of the PDF spec. Should not this flag be considered as deprecated only for PDF 2.0 files?
I analysed a corpus of about 2000 files and compared the results I had from pqdf with those of the Acrobat Pro accessibility checker. qpdf always read this flag reliably, except for the file I mentioned in this issue.

**Comment by jberkenbilt** [comment]
So I just looked at the qpdf code. Contrary to my previous comment, it always clears bit 10. The value of /P in your file is -1836, which is `1111100011010011`. In the PDF spec, the low-order bit is 1, so in this case, bit 5 is set and bit 10 is clear. With bit 10 clear, qpdf reports that extraction for accessibility is not allowed. However, bit 5 being set indicates that extraction is allowed "for any purpose". It seems that the combination of bit 5 set and bit 10 clear is a bit of a gray area. I'm not surprised you're seeing inconsistent results. It would be interesting to see the values of bits 5 and 10 in your corpus of files. I wouldn't be surprised if most of them have either both bits set, both bits clear, or bit 10 set and bit 5 clear rather than being like this one with bit 10 clear and bit 5 set. I used this shell fragment to look at the bits: `echo 65536 1836 - 2 o p | dc`

Regarding how you should treat this, I think you'd be justified in only ignoring bit 10 for PDF 2.0, but there aren't very many pdf 2.0 files in the wild since writers tend to set the header version to as low a number as they can get away with for maximum viewer compatibility. Your file is PDF 1.4.

My personal opinion is that all readers should act as if bit 10 is set, regardless of whether it is set or not, but that's just my opinion. In other words, there's nothing about this change that has anything to do with the file format itself. It's more of a reflection that disallowing extraction for accessibility is just a wrong kind of thing to do. It's more of a social opinion than a technical one.

**Comment by AlainVagner** [comment]
Thanks for these infos, this is really helpful! I have read the PDF 1.7 spec, but I missed this info about the fifth bit.
I find the definition of bit 5 not so clear:
"(Security  handlers  of  revision  2) Copy  or  otherwise  extract  text and graphics from the document, including extracting text and   graphics   (in   support   of   accessibility   to   users   with   disabilities or for other purposes). (Security handlers of revision 3 or greater) Copy or otherwise extract  text  and  graphics  from  the  document  by  operations  **other than that controlled by bit 10**. "

In my dataset, I have 566 encrypted files, in these files, the one is mentioned is the only case where the bit 5 is set and bit 10 is not set. In Acrobat Pro, if you activate the permission associated to bit 5 "enable copying of text, images and other content" then the permission associated to bit 10 is automatically activated. It looks like bit 5 overrides bit 10. I also checked that this document was readable with a screen reader and it is the case.
To do my checks in Python, I used pikepdf to extract the value of the P field, and then with the bitstring library, I did this:

```
bits = BitArray(intbe=pdf.encryption.P, length=16)
result['bit10'] = bits[16-10]
result['bit5'] = bits[16-5]
```

As you mentioned, if the flags have the same value, there is no problem.
When investigating further the grey zone, I have found that when bit 10 is set and bit 5 is not set, pikepdf reports sometimes the accessibility permission as true (174 files) and sometimes as false (88 files), so there is another parameter at play.

Is this the method where you compute if the content could be extracted for accessibility?
https://github.com/qpdf/qpdf/blob/8971443e4680fc1c0babe56da58cc9070a9dae2e/libqpdf/QPDF_encryption.cc#L1486

From what I see here, I should also check the value of R.

I agree with your opinion that PDF readers should ignore bit 10, but for the time being, it seems that it is still enforced. So I would like to be able to detect these files, for awareness purposes.

Many thanks for your help and for all your work on this invaluable project.

**Comment by jberkenbilt** [comment]
Yes, you have found the right bit of code, and if you probably should be checking `/R` as well for the reasons you discovered by looking at the code. The spec is pretty clear about what the fields mean with different values of `/R`. There's `/R` and `/V` which both determine certain aspects of how the security handler works. Not all combinations are `/R` and `/V` are valid. I can't keep that stuff in my head. I have to check the spec or read my own code to remember how it all works. It's quite arcane. There are parts of the spec that are actually ambiguous. I had to do quite a lot of trial and error and processing of lots of files generated by Acrobat to get it right, something akin to what you seem to be doing now. Basically there are some changes where they bumped `/R`, some where they bumped `/V`, and some where they bumped both. As far as I recall, only `/R` is needed to interpret `/P`, but I might be remembering that wrong. The 1.7 spec should be adequate for understanding this. By my recollection, the only change for that in 2.0 was the deprecation of disallowing accessibility.

All that said, pikepdf is just reporting what qpdf says, and qpdf is just reporting how I interpreted the spec. As you can see, different readers do different things with it. No one has ever reported a bug in how qpdf does this, but you are probably looking at this more deeply than anyone else I've discussed it with. It doesn't surprise me at all that you're seeing this kind of inconsistent behavior across readers. I see this every time I dig deep into something, often finding bugs in multiple readers. I am a big fan of PDF (not surprisingly), but it is definitely a format that shows its age and has lots of rough edges just because of the amount of evolution it has had with the strong commitment to backward compatibility. Actually qpdf suffers from some of that too. It's inevitable over time. Parts of the qpdf codebase are pushing 20 years old, and encryption was one of the first things I implemented long before qpdf was open source. In fact, the reason I initially created qpdf was that I needed something that could write linearized encrypted files, and I couldn't find anything affordable that could do that at the time.

**Comment by AlainVagner** [comment]
Thanks for the infos, I was able to test my data as you suggested with those extra fields.
The case I encountered here is indeed really rare. On the 24000 files I crawled, I encountered it only once. 
I put some logs in my code to see if I encounter it again with other values of /R and /V.
As this case is extremely rare and from what I see, you implemented the specification properly, I propose to close this ticket.
Many thanks for your help in understanding this issue.
